lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -3013,4 +3013,73 @@ public void onPresenceMessage(PresenceMessage messages) {
 				ably2.close();
 		}
 	}
+
+	/**
+	 * Authenticate using wildcard token, initialize AblyRealtime so clientId is not known a priori,
+	 * call enter() without attaching first, start connection
+	 *
+	 * Expect NACK from the server because client is unidentified
+	 *
+	 * Tests RTP8i, RTP8f, partial tests for RTP9e, RTP10e
+	 */
+	@Test
+	public void enter_before_clientid_is_known() throws AblyException {
+		AblyRealtime ably = null;
+		try {
+			ClientOptions restOpts = createOptions(testVars.keys[0].keyStr);
+			AblyRest ablyForToken = new AblyRest(restOpts);
+
+			/* Initialize connection so clientId is not known before actual connection */
+			Auth.TokenParams tokenParams = new Auth.TokenParams();
+			Capability capability = new Capability();
+			tokenParams.capability = capability.toString();
+			tokenParams.clientId = "*";
+
+			Auth.TokenDetails token = ablyForToken.auth.requestToken(tokenParams, null);
+			assertNotNull("Expected token value", token.token);
+
+			ClientOptions opts = createOptions();
+			opts.defaultTokenParams.clientId = "*";
+			opts.token = token.token;
+			opts.autoConnect = false;
+			ably = new AblyRealtime(opts);
+
+			/* enter without attaching first */
+			Channel channel = ably.channels.get("enter_before_clientid_is_known"+testParams.name);
+			CompletionWaiter completionWaiter = new CompletionWaiter();
+			channel.presence.enter(null, completionWaiter);
+
+			ably.connection.connect();
+
+			completionWaiter.waitFor(1);
+			assertFalse("Verify enter() failed", completionWaiter.success);
```

Comment:
I think the test should get the error from the CompletionWaiter and verify that it has the expected error code